Generator improvements for external modules

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

The point of this PR is to allow external modules to run `bundle exec rake decidim:generate_external_development_app` to generate a development application pointing to the module being developed, just like we do `bundle exec rake development_app` in core to generate a development application pointing to our local copy of `decidim`.

In the process of adding and testing this I discovered several issues with generator specs that I fixed. In particular,

* Some warnings and incorrect dependencies in decidim-debates `gemspec`.
* Some improvements in the generated application Gemfile: semver dependency specificators, and some unnecessary dependencies removed.
* Some other fixes in the specs related to `bundler` and `rubgems` low level details.

#### :pushpin: Related Issues
_None_.

#### :clipboard: Subtasks
_None_.

### :camera: Screenshots (optional)
_None_.

mrcasals commented on decidim-debates/decidim-debates.gemspec:
```diff
@@ -6,17 +6,20 @@ require "decidim/debates/version"
 
-  s.add_dependency "decidim-comments"
-  s.add_dependency "decidim-core"
+  s.add_dependency "decidim-comments", Decidim::Debates.version
```
Whoops, that was me when importing the code from the external module, thanks for noticing!

mrcasals commented:
I see the generator tests are fixed and the output is much better now, compare current master and this PR tests:

https://circleci.com/gh/decidim/decidim/52303 (master)
https://circleci.com/gh/decidim/decidim/52280 (this PR)

In this PR, though, I see this error message:

```
Invalid application name generator_test_app, constant GeneratorTestApp is already in use. Please choose another application name.
```

Any idea what's causing it? I don't see any test checking this case, so I don't know what's causing it 😕

deivid-rodriguez commented:
@mrcasals I have an idea, I'll have a look!

deivid-rodriguez commented:
@mrcasals So the current generator tests are not working very good. All the errors that you pointed out that were happening before should've made them fail, but didn't.

In this case, it's the same situation, the error is a legit one but it's not making the specs fail. The reason is that the spec tries to generate a development application from a decidim application (not core, or a module), and that's not expected to work nor makes sense.

So we have to separate problems:

* Generator specs don't fail when they should, and that's the only thing that they are supposed to be checking :rofl:.

* The test for `decidim:generate_external_development_application` that I added is plain wrong.

Unfortunately I can't devote time to investigate either of those, so I propose that I remove the bad test and move on for now, since this seems an improvement anyways. Thoughts?

mrcasals commented:
The workaround sounds good to me. Can you add some comments in the generator specs file to note these issues?

deivid-rodriguez commented:
Yes, definitely!

deivid-rodriguez commented:
I added a TODO comment but codeclimate is not happy. Should I open a ticket instead?

deivid-rodriguez commented:
Ok, so:

* I added another simplification here. In order to generate an application to test `decidim-debates` in `decidim.barcelona`, we were tweaking some `app_const_base` setting to be able to generate a test application for `decidim-debates` from inside the root `decidim.barcelona`. However, I think that's innecessary and switching to the plugin folder and generating the test application there would've sufficed. So I removed that setting. That's irrelevant now anyways since `decidim-debates` has been moved to core. Tha'ts https://github.com/decidim/decidim/pull/2699/commits/730a9bfbd1d3d00b706e1f5fc456547df7f838bd.

* Other than that, I discovered the problem with the generator specs. My bet is that it comes down to a bug in `thor` (which generators use under the hood):

```
$ bundle exec rake decidim:generate_external_development_app
Invalid application name decidim_barcelona, constant DecidimBarcelona is already in use. Please choose another application name.
$ echo $?
0
```

The solution is again to remove the test since you now can't (nor makes sense) to generate development or test applications from inside end applications. And since these tasks now exactly mimic `rake test_app ` and `rake development_app` and we know those work because we use them all the time, I don't think we need to test those.

deivid-rodriguez commented:
By the way, the bug is most likely https://github.com/erikhuda/thor/issues/244, which has two open PR's fixing it.

mrcasals commented:
> I added a TODO comment but codeclimate is not happy. Should I open a ticket instead?

I see no CodeClimate errors now, although I would have been OK leaving the TODO comment there.

Apart from that, awesome PR, as always! 🎉

deivid-rodriguez commented:
Thanks @mrcasals, yeah I ended up fixing the problem so no TODO comment was needed! :)
